Allow multi-dimensional dirichet (correct pull)

MichielCottaar opened the pull request:
This alters the StickBreaking transformation to allow for multi-dimensional Dirichlet object. This should resolve #792 . The first dimension is treated as special and the sum along this dimension of the resultant variables are consistenlty one. Both the log(p) and the jacobian of the transformation are now multi-dimensional array with summation only over the first dimension, which means that a useful logp_elemwiset is created.

MichielCottaar committed:
BUG allow Dirichlet to be multi-dimensional

The resulting variables summed over the first axis will be one.

MichielCottaar committed:
re-added summing over first dimension to fix test

jsalvatier commented on pymc3/distributions/transforms.py:
```diff
@@ -150,32 +150,32 @@ def __init__(self):
         z = x0/s
         Km1 = x.shape[0] - 1
-        k = arange(Km1)
+        k = arange(Km1)[(slice(None), ) + (None, ) * (x.ndim - 1)]
```
What exactly is this doing? Might be good to leave a note

MichielCottaar replied:
This ensures that k (and hence eq_share) has the same number of dimensions as x, which is necessary for it to be subtracted from logit(z).

jsalvatier commented:
Thanks for this Michiel! Super helpful. 

Looks good to me. Wan we also add a test for 2-d variables?

MichielCottaar commented:
The most obvious test to add seems to me to be to compare the log(p) of a dirichlet of shape (3, 5) with 5 individual dirichlets of shape (3,) and ensure that they are the same (where 3 and 5 are just example numbers).
However, I don't see an obvious way to extend the test_dirichlet in test_distributions to do this. Do you have any suggestions, where I could add a new test for this or how I could extend the test_dirichlet to test a multi-dimensional case.

jsalvatier commented:
Since most of the change was in the transform, I would add a test in https://github.com/pymc-devs/pymc3/blob/master/pymc3/tests/test_transforms.py. 

Maybe just add a line to https://github.com/pymc-devs/pymc3/blob/master/pymc3/tests/test_transforms.py#L30

Where you pass in a different `Domain`, instead of `Simplex` make new "multi-simplex" space that has a bunch simplexes. The simplex domain is defined here: https://github.com/pymc-devs/pymc3/blob/master/pymc3/tests/test_distributions.py#L88

You can probably copy this and modify it slightly. You can also look at the other domains defined nearby such as `PdMatrix` or `R`.

twiecki commented:
@MichielCottaar Any updates on this?

MichielCottaar committed:
Added test for 2D stick breaking transformation

MichielCottaar committed:
tested log(p) of 2D dirichlet

The main goal of this test is to show that the variables are independent
across the second dimension.
Note that betafn and dirichlet_logpdf had to be changed so that
internally they only sum over the first dimension, and only in the final
summation is the log(p) summed over all dimensions.

MichielCottaar commented:
I've added both the test suggested by John, as well as a test to compute the log(p) of a 2D dirichlet. Unfortunately the second test takes about a minute (mainly because the parameter space explored is big).

twiecki commented:
Great, thanks @MichielCottaar! I'm not worried about the length of the test.

twiecki commented:
Merged with https://github.com/pymc-devs/pymc3/commit/4a62e8c28c51df4a680ef8933a6b022c64894f73.
